drivers/lutron/vive_leap.cr

```diff
@@ -0,0 +1,253 @@
+  def disconnected
+    @sensors.clear
+    schedule.clear
+  end
+
+  def connected
+    # this request needs to be made before anything else to negotiate protocol version
+    request = Request.new("/clientsetting", :update_request, {
+      ClientSetting: {
+        ClientMajorVersion: 1
+      }
+    })
+    send request.to_json, priority: 99, name: request.name?
+
+    schedule.every(1.minute) { ping }
+  end
+
+  # This is called after the protocol version is negotiated
+  protected def authenticate
+    request = Request.new("/login", :update_request, {
+      Login: {
+        ContextType: "Application",
+        LoginId: @username,
+        Password: @password
+      }
+    })
+    send request.to_json, priority: 99, name: request.name?
+  end
+
+  def ping
+    request = Request.new("/server/status/ping")
+    send request.to_json, priority: 0, name: request.name?
+  end
+
+  # gets the status of all areas
+  def area_status?
+    request = Request.new("/area/status")
+    send request.to_json, name: request.name?
+  end
+
+  protected def subscribe_areas
+    request = Request.new("/area/status", :subscribe_request)
+    send request.to_json, name: :subscribe_area_status
+  end
+
+  # get the status of all zones
+  def zone_status?
+    request = Request.new("/zone/status")
+    send request.to_json, name: request.name?
+  end
+
+  protected def subscribe_zones
+    request = Request.new("/zone/status", :subscribe_request)
+    send request.to_json, name: :subscribe_zone_status
+  end
+
+  def zone_level(zone_id : String | Int32, level : Float64)
+    request = Request.new("/zone/#{zone_id}/commandprocessor", :create_request, {
+      Command: {
+        CommandType: "GoToDimmedLevel",
+        DimmedLevelParameters: {
+          Level: level
+        }
+      }
+    })
+    send request.to_json, name: request.name?
+  end
+
+  def zone_lighting(zone_id : String | Int32, state : Bool)
+    request = Request.new("/zone/#{zone_id}/commandprocessor", :create_request, {
+      Command: {
+        CommandType: "GoToSwitchedLevel",
+        SwitchedLevelParameters: {
+          SwitchedLevel: state ? "On" : "Off"
+        }
+      }
+    })
+    send request.to_json, name: request.name?
+  end
+
+  def zone_contact_closure(zone_id : String | Int32, state : Bool)
+    request = Request.new("/zone/#{zone_id}/commandprocessor", :create_request, {
+      Command: {
+        CommandType: "GoToCCOLevel",
+        CCOLevelParameters: {
+          CCOLevel: state ? "Closed" : "Open"
+        }
+      }
+    })
+    send request.to_json, name: request.name?
+  end
+
+  def received(data, task)
+    data = String.new(data)
+    logger.debug { "Lutron sent: #{data}" }
+    request = Request.from_json(data)
+
+    url = request["Url"]?
+    status = request["StatusCode"]? || "200 OK"
+    message_type = request["MessageBodyType"]?
+
+    # process the message based on its type by preference
+    case message_type
+    when "OneClientSettingDefinition"
```

Comment:
Maybe make an enum for the MessageBodyTypes so you can do `case in...` to ensure exhaustive coverage?

Reply:
Style spec is also failing

Reply:
I personally don't think it's required here as we only care about a small subset of the message body types